2021-year-report.org

```diff
@@ -0,0 +1,438 @@
+  fn bar(self) {
+    println!("In trait impl!")
+  }
+}
+
+fn main() {
+  let mut f = Foo{};
+  f.bar();
+}
+#+END_SRC
+
+This test case is all about showing even though there are two duplicate methods here, we enforce that the impl block method is prefered first.
+
+see:
+https://rustc-dev-guide.rust-lang.org/method-lookup.html
+https://doc.rust-lang.org/nightly/nomicon/dot-operator.html
+
+**** Add check for duplicate overlapping impl-items
+
+Rust allows multiple impl blocks to give a generic data type specialization. But suppose the programmer adds a generic impl for a duplicate method. In that case, it will become impossible to distinguish the method 'bar' in each of these specialized impl blocks for method resolution. Since your receiver could be Foo<_ (inference variable)> which could resolve to Foo<isize> or Foo<char>. see: rustc --explain E0592
+
+#+BEGIN_SRC rust
+struct Foo<A>(A);
+
+impl Foo<isize> {
+    fn bar(self) -> isize {
+        self.0
+    }
+}
+
+impl Foo<char> {
+    fn bar(self) -> char {
+        self.0
+    }
+}
+
+impl<T> Foo<T> {
+    fn bar(self) -> T {
+        self.0
+    }
+}
+#+END_SRC
+
+**** Support Dereference operator overloading
+
+Deref operator overloading is a core piece of Rusts control flow mechanism, it adds in support for more complex method resolution cases as part of the autoderef mechanism. It also has served as a good test of the current state of the type system so far. 
+
+#+BEGIN_SRC rust
+extern "C" {
+    fn printf(s: *const i8, ...);
+}
+
+#[lang = "deref"]
+pub trait Deref {
+    type Target;
+
+    fn deref(&self) -> &Self::Target;
+}
+
+impl<T> Deref for &T {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        *self
+    }
+}
+
+impl<T> Deref for &mut T {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        *self
+    }
+}
+
+struct Foo<T>(T);
+impl<T> Deref for Foo<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+fn main() -> i32 {
+    let foo: Foo<i32> = Foo(123);
+    let bar: i32 = *foo;
+
+    unsafe {
+        let a = "%i\n\0";
+        let b = a as *const str;
+        let c = b as *const i8;
+
+        printf(c, bar);
+    }
+
+    0
+}
+#+END_SRC
+
+The interesting piece about dereferences is that the actual deref method that is implemented _always_ returns a reference to the associated type 'Target', this implicitly requires the compiler call this method and because the trait and type checking ensures that the result is a reference it means it can safely be dereferenced by the compiler implicitly. I point this out because simply because the function prototype:
+
+#+BEGIN_SRC rust
+fn deref(&self) -> &Self::Target {
+    &self.0
+}
+#+END_SRC
+
+Here the function type is: 
+
+#+BEGIN_SRC rust
+fn deref(self: &Foo<T>) -> &T { &self.0 }
+#+END_SRC
+
+So the dereference operation even on custom types is always going to return a reference. So the dereference operator overloading is a two step mechanism.
```

Comment:
Same here. What was challenging about this? Where is documentation missing for this? Also this section is weirdly worded.